### PR TITLE
Start ProtoArray from its own snapshot

### LINKED ISSUE #2160: Avoid regenerating all states to initialise protoarray
### Description
At startup we regenerate all states and then ProtoArray walks through every block in the store and also requests the states for them.  When states have been pruned this leads to a lot of extra regeneration and slow startup times.

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Starts ProtoArray from its own snapshot. Save protoArray snapshot to storage every epoch. Tested manually to confirm no issues arise.

## Fixed Issue(s)

Fixes https://github.com/PegaSysEng/teku/issues/2160

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArrayForkChoiceStrategy.java [hunk 1]
```diff
@@ -123,7 +145,11 @@ public void maybePrune(Bytes32 finalizedRoot) {
 
   // Internal
   private static void processBlocksInStoreAtStartup(ReadOnlyStore store, ProtoArray protoArray) {
+    List<Bytes32> alreadyIncludedBlockRoots =
+        protoArray.getNodes().stream().map(ProtoNode::getBlockRoot).collect(Collectors.toList());
+
     store.getBlockRoots().stream()
```
Comment (mbaxter) [inline protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArrayForkChoiceStrategy.java:151]: We should be able to optimize this more by adding an API like `store.processBlocks(Consumer<SignedBlockAndState> consumer, Set<Bytes32> blockRoots)`, but we can save that for a follow-up.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 2]
```diff
@@ -93,6 +93,10 @@ public AttestationProcessingResult onAttestation(
     return on_attestation(store, attestation, stateTransition, getForkChoiceStrategy());
   }
 
+  public void saveToStorage() {
```
Comment (mbaxter) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]:
```suggestion
  public void save() {
```

**Review by mbaxter — APPROVED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java (hunk 2 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: Done.

**Pull request merged**